subsys/usb/device_next/usbd_core.c

```diff
@@ -134,71 +150,72 @@ static int event_handler_bus_reset(struct usbd_context *const uds_ctx)
 
 
 static ALWAYS_INLINE void usbd_event_handler(struct usbd_context *const uds_ctx,
-					     struct udc_event *const event)
+					     const uint32_t event)
 {
 	int err = 0;
 
-	switch (event->type) {
-	case UDC_EVT_VBUS_REMOVED:
+	if (event == BIT(UDC_EVT_VBUS_REMOVED)) {
 		LOG_DBG("VBUS remove event");
 		usbd_msg_pub_simple(uds_ctx, USBD_MSG_VBUS_REMOVED, 0);
-		break;
-	case UDC_EVT_VBUS_READY:
+	}
+
+	if (event == BIT(UDC_EVT_VBUS_READY)) {
 		LOG_DBG("VBUS detected event");
 		usbd_msg_pub_simple(uds_ctx, USBD_MSG_VBUS_READY, 0);
-		break;
-	case UDC_EVT_SUSPEND:
+	}
+
+	if (event == BIT(UDC_EVT_RESET)) {
+		LOG_DBG("RESET event");
+		err = event_handler_bus_reset(uds_ctx);
+		usbd_msg_pub_simple(uds_ctx, USBD_MSG_RESET, 0);
+	}
+
+	if (event == BIT(UDC_EVT_SUSPEND)) {
 		LOG_DBG("SUSPEND event");
 		usbd_status_suspended(uds_ctx, true);
-		usbd_class_bcast_event(uds_ctx, event);
+		usbd_class_bcast_event(uds_ctx, UDC_EVT_SUSPEND);
 		usbd_msg_pub_simple(uds_ctx, USBD_MSG_SUSPEND, 0);
-		break;
-	case UDC_EVT_RESUME:
+	}
+
+	if (event == BIT(UDC_EVT_RESUME)) {
 		LOG_DBG("RESUME event");
 		usbd_status_suspended(uds_ctx, false);
-		usbd_class_bcast_event(uds_ctx, event);
+		usbd_class_bcast_event(uds_ctx, UDC_EVT_RESUME);
 		usbd_msg_pub_simple(uds_ctx, USBD_MSG_RESUME, 0);
-		break;
-	case UDC_EVT_SOF:
-		usbd_class_bcast_event(uds_ctx, event);
-		break;
-	case UDC_EVT_RESET:
-		LOG_DBG("RESET event");
-		err = event_handler_bus_reset(uds_ctx);
-		usbd_msg_pub_simple(uds_ctx, USBD_MSG_RESET, 0);
-		break;
-	case UDC_EVT_EP_REQUEST:
-		err = event_handler_ep_request(uds_ctx, event);
-		break;
-	case UDC_EVT_ERROR:
+	}
+
+	if (event == BIT(UDC_EVT_SOF)) {
+		usbd_class_bcast_event(uds_ctx, UDC_EVT_SOF);
+	}
+
+	if (event == BIT(UDC_EVT_ERROR)) {
 		LOG_ERR("UDC error event");
-		usbd_msg_pub_simple(uds_ctx, USBD_MSG_UDC_ERROR, event->status);
-		break;
-	default:
-		break;
-	};
+		usbd_msg_pub_simple(uds_ctx, USBD_MSG_UDC_ERROR, 0);
+	}
 
 	if (err) {
 		usbd_msg_pub_simple(uds_ctx, USBD_MSG_STACK_ERROR, err);
 	}
 }
 
-static void usbd_thread(void *p1, void *p2, void *p3)
+static void usbd_thread(void *const p1, void *p2, void *p3)
 {
-	ARG_UNUSED(p1);
 	ARG_UNUSED(p2);
 	ARG_UNUSED(p3);
-
-	struct usbd_context *uds_ctx;
-	struct udc_event event;
+	struct usbd_context *const uds_ctx = p1;
 
 	while (true) {
-		k_msgq_get(&usbd_msgq, &event, K_FOREVER);
+		uint32_t events;
+
+		events = k_event_wait(&uds_ctx->events, UINT32_MAX, false, K_FOREVER);
```

Comment:
What happens if multiple events come? The msgq was going to keep them, this will seemingly lose duplicates?

Reply:
Well, yes, that could be an issue, that is why it is a draft. We also lose some metadata from the events. Losing some duplicates of the bus event may be intended.